### PR TITLE
Add cuSPARSELt

### PR DESCRIPTION
**Pull request by mnicely**
Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [X] Source is from official source.
- [X] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Build number is 0.
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

@leofang If you have a chance, can you double check my work?

### CODE REVIEW

File: recipes/cusparselt/conda_build_config.yaml
Comment (leofang) [inline recipes/cusparselt/conda_build_config.yaml:18]: Please check out https://github.com/conda-forge/cutensor-feedstock/blob/master/recipe/conda_build_config.yaml. Conda-Forge zips 4 keys together: `cudnn`, `cuda_compiler_version`, `docker_image` and `cdt_name`. So if we are implementing a custom `conda_build_config.yaml` to overwrite the default setting in `conda-forge-pinning-feedstock`, we nee to be explicit on allowed keys.

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:23]:
```suggestion
  skip: True  # [not linux64 or cuda_compiler_version in ("None", "9.2", "10.0")]
```

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:28]: Does this line work for CUDA 11.0?

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:39]: This allows the package for CUDA 11.0 to work with all 11.x:
```suggestion
    - {{ compiler('c') }}    # [linux]
    - {{ compiler('cuda') }}  # [cuda_compiler_version == "11.0"]
```

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:55]: Ditto:
```suggestion
    - libstdcxx-ng >=3.4  # [linux]
    - cudatoolkit {{ cuda_compiler_version }}  # [cuda_compiler_version in ("10.1", "10.2")]
    - cudatoolkit 11.*                         # [cuda_compiler_version not in ("10.1", "10.2")]
```

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:93]: I have been helping maintain several NVIDIA packages, so I don't mind another 🙂
```suggestion
    - mnicely
    - leofang
```

**Review by leofang — COMMENTED** [review]
Thanks, @mnicely. Comments below are mostly copied from cuTENSOR/cuDNN/NCCL.

cc: @kkraus14 @jakirkham

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:41]: One more thing: From the version it looks like cusparseLt is not yet stable (for backward compatibility etc) enough to obey versioning semantics? If it's the case, then it's best to pin it exactly:  
```suggestion
    - {{ pin_subpackage('cusparseLt', exact=True) }}
```

File: recipes/cusparselt/meta.yaml
Comment (mnicely) [reply recipes/cusparselt/meta.yaml:28]: cuSPARSELt requires CUDA 11 or greater.

Comment (mnicely) [reply recipes/cusparselt/meta.yaml:28]: But yes it does. It doesn't have the same folder structure as cutensor

File: recipes/cusparselt/meta.yaml
Comment (mnicely) [reply recipes/cusparselt/meta.yaml:41]: okay!

File: recipes/cusparselt/conda_build_config.yaml
Comment (mnicely) [reply recipes/cusparselt/conda_build_config.yaml:18]: In the case of cuSPARSELt, it only works with 11.x, and the next release will require 11.2. 
Also, I'm guessing we don't need `docker_image`, `cudnn`, and `cdt_name`???

Comment (leofang) [reply recipes/cusparselt/conda_build_config.yaml:18]: It is fine leaving all available CUDA versions here, as the actual skip is done in `meta.yaml`. But whatever `docker_image` we list here, they must be accompanied by the rest of the 3 keys. The key-zipping is done globally, not per feedstock, so it is not something controllable here. This was why you got this error in Azure:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=291111&view=logs&j=232b7f5e-8d64-5438-fada-8478f379d226&t=e8516d7e-1153-5090-5807-53ae833fddd8&l=557

File: recipes/cusparselt/meta.yaml
Comment (leofang) [reply recipes/cusparselt/meta.yaml:28]: Thanks, Matt, you got what I was asking 🙂

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:23]: Sorry, I didn't know this is 11.0+ only. In this case, we need to expand the skip condition. But, since we are using 11.0 to support all 11.x, it's actually simpler:
```suggestion
  skip: True  # [not linux or cuda_compiler_version != "11.0")]
```

**Review by leofang — CHANGES_REQUESTED** [review]

File: recipes/cusparselt/conda_build_config.yaml
Comment (mnicely) [reply recipes/cusparselt/conda_build_config.yaml:18]: What "keys" are you referring to??

Comment (leofang) [reply recipes/cusparselt/conda_build_config.yaml:18]: By keys I am referring to
> `cudnn`, `cuda_compiler_version`, `docker_image` and `cdt_name`

and their values follow with a `-` prefix, say,
```
cuda_compiler_version:
  - None
  - 9.2
  - 10.0
  - ...
``` 
This is what's done in conda-forge-pinning-feedstock:
https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/a365475270cfefdd44975873091d6ad7503b7e9a/recipe/conda_build_config.yaml#L141-L145
which affects globally (i.e. all Conda-Forge packages). 

If any of these keys appear in `conda_build_config.yaml` (which we normally don't need, as the default is to inherit the setting from CF pinning), they must appear together so that they can be zipped correctly by conda-smithy. This is why this file in cuTENSOR/cuDNN/NCCL looks so complicated. I think it'd work if you just copy this file from the cuTENSOR feedstock, to which I linked earlier.

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:73]: Learned this from @beckermr to eliminate the license linter error 🙂
```suggestion
  license: LicenseRef-cuSPARSELt-Software-License-Agreement
```
We probably should do this in the next run of updates for all NVIDIA packages.

Ref: https://conda-forge.org/docs/maintainer/adding_pkgs.html?highlight=spdx#spdx-identifiers-and-expressions

Comment (kkraus14) [reply recipes/cusparselt/meta.yaml:73]: How does this affect how the license is displayed on the Anaconda web server UI?

Comment (leofang) [reply recipes/cusparselt/meta.yaml:73]: Good question, check this out:
https://anaconda.org/conda-forge/hdf5
They use `LicenseRef-HDF5`.

Comment (jakirkham) [reply recipes/cusparselt/meta.yaml:73]: It wouldn't affect the `license_url`'s usage if that's what you are referring to. Think that may come down to some issue with how packages get copied as discussed in issue ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/2 )

Comment (kkraus14) [reply recipes/cusparselt/meta.yaml:73]: Looks good. Just want to prevent NVIDIA legal from causing problems :)

File: recipes/cusparselt/conda_build_config.yaml
Comment (jakirkham) [reply recipes/cusparselt/conda_build_config.yaml:18]: Yeah I think we can drop this file. Though please correct me if I'm missing something

Comment (leofang) [reply recipes/cusparselt/conda_build_config.yaml:18]: I think that'd also work!

Comment (mnicely) [reply recipes/cusparselt/conda_build_config.yaml:18]: Okay, so I copied the file from cuTENSOR feedstock but removed all references to ppc64le. We don't plan to support Power9

Comment (leofang) [reply recipes/cusparselt/conda_build_config.yaml:18]: Matt, the 3 keys are still missing. Let's try just removing this file entirely as John suggested.

File: recipes/cusparselt/meta.yaml
Comment (leofang) [inline recipes/cusparselt/meta.yaml:24]: Typo:
```suggestion
  skip: True  # [not linux or cuda_compiler_version != "11.0"]
```

**Review by leofang — CHANGES_REQUESTED** [review]
In addition to the still incorrect `conda_build_config.yaml`, there's also one typo that caused the CIs to fail.

One more thing: @mnicely Is NVIDIA Legal ok with this? I remember it took a while to get approval for releasing cuTENSOR. Maybe you've secured it, Matt? If not, I am sure @kkraus14 can help with this 🙂

**Comment by kkraus14** [comment]
> One more thing: @mnicely Is NVIDIA Legal ok with this? I remember it took a while to get approval for releasing cuTENSOR. Maybe you've secured it, Matt? If not, I am sure @kkraus14 can help with this 🙂

In order to get approval / the pieces in place for cuTENSOR I worked with @mnicely internally 😄

**Comment by leofang** [comment]
Ahh ok, thanks Keith. Good to know that. Forgive my ignorance, I don't know how things work inside NVIDIA 😅

File: recipes/cusparselt/conda_build_config.yaml
Comment (jakirkham) [reply recipes/cusparselt/conda_build_config.yaml:18]: Thanks for the context Matt. Yeah in that case we don't need the file. Could you please drop it?

Comment (leofang) [reply recipes/cusparselt/conda_build_config.yaml:18]: Adding a bit more context in the hope that Matt wouldn't get confused by me 😄 

We have this file `conda_build_config.yaml` in other feedstocks because 1. we wanted to span the build matrix beyond the default setting in conda-forge-pinning, and 2. for certain cases we want to use CentOS 7 docker images, mostly due to the requirement of using a newer glibc version in some packages (such as cuTENSOR).

In the case of cuSPARSELt, if we're only building on CUDA 11.0 and supporting 11.x, none of the above applies. On Conda-Forge (and likely elsewhere), 11.0 is enabled by default, and all CUDA 11.x images are already on cos7.

Comment (mnicely) [reply recipes/cusparselt/conda_build_config.yaml:18]: Thanks @leofang @jakirkham! I'll go ahead and remove the file

**Comment by leofang** [comment]
@jakirkham The CI looks a bit strange -- why there isn't a `linux_64_cuda_110` entry? Is it because the migration is not over yet, and it has to wait until a feedstock is created?

File: recipes/cusparselt/meta.yaml
Comment (kkraus14) [inline recipes/cusparselt/meta.yaml:79]: Duplicated key here?

Comment (mnicely) [reply recipes/cusparselt/meta.yaml]: Good catch! 
It's been removed.

**Comment by leofang** [comment]
> The CI looks a bit strange -- why there isn't a `linux_64_cuda_110` entry? Is it because the migration is not over yet, and it has to wait until a feedstock is created?

OK I think I got it. CUDA 10.2 is the default in stage-recipes:
https://github.com/conda-forge/staged-recipes/blob/916365f20d9e1bbb6d9efd2e3e1bc7674b43b341/.ci_support/linux64.yaml#L13-L15
It'd be overwritten if a custom `conda_build_config.yaml` is provided, but since we took it out, the CI setting falls back to the default, which is fine because by the time the feedstock is created, the new settings from conda-forge-pinning will kick in. Nothing to worry.

**Review by leofang — APPROVED** [review]
LGTM!

File: recipes/cusparselt/meta.yaml
Comment (jakirkham) [inline recipes/cusparselt/meta.yaml:23]: Looks like this was leftover. Think we can go ahead and drop it

```suggestion
```

File: recipes/cusparselt/meta.yaml
Comment (jakirkham) [inline recipes/cusparselt/meta.yaml:58]: This only builds on CUDA 11.0 based on the `skip` above. So think we can just drop this selector

```suggestion
    - cudatoolkit 11.*
```

File: recipes/cusparselt/meta.yaml
Comment (jakirkham) [inline recipes/cusparselt/meta.yaml:66]: Should this be dropped?

```suggestion
```

File: recipes/cusparselt/meta.yaml
Comment (mnicely) [reply recipes/cusparselt/meta.yaml:58]: Done

File: recipes/cusparselt/meta.yaml
Comment (mnicely) [reply recipes/cusparselt/meta.yaml:66]: Yes

File: recipes/cusparselt/post-link.sh
Comment (jakirkham) [inline recipes/cusparselt/post-link.sh:3]:
```suggestion
echo "By downloading and using the cuSPARSELt conda package, you accept the terms"    >> "$PREFIX/.messages.txt"
echo "and conditions of the NVIDIA cuSPARSELt Software License Agreement:"            >> "$PREFIX/.messages.txt"
echo "https://docs.nvidia.com/cuda/cusparselt/license.html"                           >> "$PREFIX/.messages.txt"
```

File: recipes/cusparselt/post-link.bat
Comment (jakirkham) [inline recipes/cusparselt/post-link.bat:4]:
```suggestion
echo "By downloading and using the cuSPARSELt conda package, you accept the terms"    >> "%PREFIX%\.messages.txt"
echo "and conditions of the NVIDIA cuSPARSELt Software License Agreement:"            >> "%PREFIX%\.messages.txt"
echo "https://docs.nvidia.com/cuda/cusparselt/license.html"                           >> "%PREFIX%\.messages.txt"
```

File: recipes/cusparselt/meta.yaml
Comment (jakirkham) [inline recipes/cusparselt/meta.yaml:11]:
```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Sorry missed a couple minor things before

File: recipes/cusparselt/meta.yaml
Comment (jakirkham) [reply recipes/cusparselt/meta.yaml:16]: Looks like this still needs to be dropped

**Comment by mnicely** [comment]
Thanks @jakirkham!

**Review by jakirkham — APPROVED** [review]

**Review by kkraus14 — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks Matthew for the PR and everyone for the reviews! 😄

**Comment by leofang** [comment]
Thanks Matt and everyone!

**Pull request merged**
